### PR TITLE
Implementation of surv.survdnn

### PR DESCRIPTION
**Pull request by ielbadisy**
This PR adds a minimal LearnerSurvSurvDNN wrapping survdnn::survdnn().

- Supports prediction types: crank, lp, distr
- Compatible with mlr3 resampling and mlr3tuning
- Early stopping and internal tuning intentionally left out for a first iteration

Ref: ielbadisy/survdnn#24

### CODE REVIEW

File: R/learner_surv_survdnn.R [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        .device     = paradox::p_fct(levels = c("auto", "cpu", "cuda"), default = "auto", tags = "train"),
+        na_action   = paradox::p_fct(levels = c("omit", "fail"), default = "omit", tags = "train")
+      )
+
+      super$initialize(
+        id = "surv.survdnn",
+        packages = c("mlr3extralearners", "survdnn"),
+        feature_types = c("integer", "numeric", "factor", "ordered"),
+        predict_types = c("crank", "lp", "distr"),
+        properties = c("weights", "missings"),
```
Comment (bblodfon) [inline R/learner_surv_survdnn.R]: see https://mlr3.mlr-org.com/reference/Learner.html#weights, do you support this? `missings` means you can train with missing features, is that true? check available properties here => https://mlr3.mlr-org.com/reference/Learner.html#method-new-

File: R/learner_surv_survdnn.R [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      # fallback no events (should be rare)
+      if (length(times) == 0L) {
+        return(mlr3proba::surv_return(crank = lp, lp = lp))
+      }
+
+      # survival matrix on time grid
+      surv_df = stats::predict(model, newdata = newdata, times = times, type = "survival")
+      surv = as.matrix(surv_df)
+
+      mlr3proba::surv_return(times = times, surv = surv, crank = lp, lp = lp)
```
Comment (bblodfon) [inline R/learner_surv_survdnn.R]: `crank` is automatically asssigned the `lp` (see https://mlr3proba.mlr-org.com/reference/surv_return.html), so you don't necesserily need to add it!

File: R/learner_surv_survdnn.R [hunk 3]
```diff
@@ -0,0 +1,95 @@
+#' @title SurvDNN Survival Learner
+#' @name mlr_learners_surv.survdnn
+#' @author ielbadisy
+#'
+#' @description
+#' Deep neural network survival models from package \CRANpkg{survdnn}.
+#' Wraps [survdnn::survdnn()] and [stats::predict()] for class `"survdnn"`.
+#'
+#' @section Prediction types:
```
[the remark this answers is not in the document]
Comment (ielbadisy) [reply R/learner_surv_survdnn.R]: Documentation explicitly extended to describe the prediction types, baseline estimation, and the use of training event times.

File: R/learner_surv_survdnn.R [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#' @template learner
+#' @templateVar id surv.survdnn
+#' @export
+LearnerSurvSurvDNN = R6::R6Class("LearnerSurvSurvDNN",
+  inherit = mlr3proba::LearnerSurv,
+
+  public = list(
+    initialize = function() {
+
+      ps = paradox::ps(
```
[the remark this answers is not in the document]
Comment (ielbadisy) [reply R/learner_surv_survdnn.R]: renamed to `param_set`

File: R/learner_surv_survdnn.R [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        hidden      = paradox::p_uty(tags = "train", default = c(32L, 16L)),
+        activation  = paradox::p_fct(
+          levels = c("relu", "leaky_relu", "tanh", "sigmoid", "gelu", "elu", "softplus"),
+          default = "relu",
+          tags = "train"
+        ),
+        lr          = paradox::p_dbl(lower = 1e-6, upper = 1, default = 1e-4, tags = "train"),
+        epochs      = paradox::p_int(lower = 1L, default = 300L, tags = "train"),
+        loss        = paradox::p_fct(levels = c("cox", "cox_l2", "aft", "coxtime"), default = "cox", tags = "train"),
```
[the remark this answers is not in the document]
Comment (ielbadisy) [reply R/learner_surv_survdnn.R]: Good point. I’m currently working on this mapping in a paper under review at The R Journal. I’ll add it here once that is finalized.

The package supports both low and high-dimensional tabular data.
.

File: R/learner_surv_survdnn.R [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        optimizer   = paradox::p_fct(levels = c("adam", "adamw", "sgd", "rmsprop", "adagrad"), default = "adam", tags = "train"),
+        optim_args  = paradox::p_uty(default = list(), tags = "train"),
+        verbose     = paradox::p_lgl(default = FALSE, tags = "train"),
+        dropout     = paradox::p_dbl(lower = 0, upper = 1, default = 0.3, tags = "train"),
+        batch_norm  = paradox::p_lgl(default = TRUE, tags = "train"),
+        callbacks   = paradox::p_uty(default = NULL, tags = "train"),
+        .seed       = paradox::p_int(default = NULL, special_vals = list(NULL), tags = "train"),
+        .device     = paradox::p_fct(levels = c("auto", "cpu", "cuda"), default = "auto", tags = "train"),
+        na_action   = paradox::p_fct(levels = c("omit", "fail"), default = "omit", tags = "train")
```
[the remark this answers is not in the document]
Comment (ielbadisy) [reply R/learner_surv_survdnn.R]: For now, i have kept the parameters set limited to user-facing training parameters and did not expose internal loss specific quantities, as these are learned model attributes rather than tuning parameters. 
I can revisit this if exposing them is considered useful from mlr3 perspective.

File: R/learner_surv_survdnn.R [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    .predict = function(task) {
+      model = self$model
+      newdata = ordered_features(task, self)
+
+      # linear predictor
+      lp = as.numeric(stats::predict(model, newdata = newdata, type = "lp"))
+
+      # build time grid from training event times
+      y_train = stats::model.response(stats::model.frame(model$formula, model$data))
```
[the remark this answers is not in the document]
Comment (ielbadisy) [reply R/learner_surv_survdnn.R]: The training `Surv` object is now stored during `.train()`.

File: R/learner_surv_survdnn.R [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      model = mlr3misc::invoke(
+        survdnn::survdnn,
+        formula = task$formula(),
+        data = task$data(),
+        .args = pv
+      )
+
+      # store training survival outcome + default time grid for predictions
+      y_train = stats::model.response(stats::model.frame(task$formula(), task$data()))
+      self$state$y_train = y_train
```
Comment (bblodfon) [inline R/learner_surv_survdnn.R]: This (`self$state`) gets erased (expect of the `self$state$model`) when you use a graph learner + possibly doing encapsulation (we have had huge issues with this before) => so my suggestion is to add the `y` train info either in the model (if a `list`) or as an `attribute` of it. This will solve it for this particular `mlr3`-case. What is even better, is to store it in the returned object of `survdnn::survdnn()` directly - always useful to have the training outcome stored, R modeling packages do it all the time!

Comment (ielbadisy) [reply R/learner_surv_survdnn.R]: `self$state` is now avoided for this. The training `Surv` outcome and the default event-time grid are now stored on the fitted `survdnn` object, as attributes.

File: tests/testthat/test_survdnn_surv_survdnn.R [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  learner = lrn("surv.survdnn",
+    loss = "cox",
+    epochs = 2L,
+    verbose = FALSE,
+    .device = "cpu",
+    .seed = 1L
+  )
+  expect_learner(learner)
+
+  result = run_autotest(
```
Comment (bblodfon) [inline tests/testthat/test_survdnn_surv_survdnn.R]: @ielbadisy have you tried running this locally? remember to run first `devtools::install()` so that when the encapsulation test happen the proper learner code is used. I think we need to implement marshaling for this learner, ie this doesn\t work : 

```
library(mlr3extralearners)

l = lrn("surv.survdnn", epochs = 42)
task = tsk("rats")
l$train(task)
saveRDS(l, "model.rds")

# reload R, try to predict
l = readRDS("model.rds")
l$predict(task)
# External pointer is not valid!
```
